### PR TITLE
Faster implementation for Free

### PR DESCRIPTION
**Pull request by thomashoneyman**
I [tested a new implementation for Free](https://github.com/thomashoneyman/purescript-halogen-hooks/pull/61) when benchmarking Halogen Hooks and found it sped up the library (and Halogen) by as much as 25% while using less memory.

This PR adds this new implementation while preserving the existing API. The benchmarks prove the new implementation is significantly more performant than the existing v5.2.0 implementation -- an even greater performance difference than between the current version and v0.6.1 (see benchmarks below).

That means Halogen, Run, and other Free-based applications will see a nice performance boost going forward.

## API Changes

This PR preserves the existing API and introduces three new functions: `run`, `interpret`, and `interpretRec`.

Of these functions, `interpret` is a drop-in replacement for `substFree` _and_ for `foldFree`, and `interpretRec` is a drop-in replacement for `foldFree` (with the `MonadRec` constraint).

I think these names better indicate what the functions are for (interpreting a Free structure into another monad or Free structure, or doing the same thing except with a `MonadRec` constraint). `substFree` and `foldFree` still exist, just with a deprecation warning. If desired, I can remove the deprecation warnings and only add the `interpret` function.

In the future I think we should adopt more of @natefaubion's implementation, which I can put in a separate issue. In my opinion it provides an API with simpler, more approachable naming conventions than those which currently exist. Specifically, it drops suffixes except for functions which have a `MonadRec` constraint, so:

`liftF` -> `lift`
`suspendF` -> `suspend`
`hoistFree` -> `hoist`
`runFree` -> `run`
`runFreeM` -> `runRec`

In addition, the `substFree` and `foldFree` functions are renamed to better clarify what they're doing:

`substFree` -> `interpret`
`foldFree` -> `interpretRec`

However, I think these changes are worth exploring in a separate proposal because they would be breaking changes.

## Notes

* Only commits from https://github.com/purescript/purescript-free/commit/3bf25faab376e864ddf80e887d37fb9fe29b0d66 onward are relevant to this PR. The prior commits are the same as #103, which just fixes the benchmarks but doesn't change any of the implementations. You might want to look at the diff with only that PR in mind.

*  ~~To avoid too many breaking changes I've added a `Compat` module for any modules which would no longer work in user code. These `Compat` modules can be directly dropped in to preserve existing behavior in libraries and applications. But I think they should be removed in a future version.~~ I have instead preserved the existing API.

## Benchmark Results

left-bind-small

(Note: v6.0.1 stack overflows on this benchmark, so we're only comparing this implementation to the current v5.2.0 implementation)
left-bind-large

right-bind-small

right-bind-large

### CODE REVIEW

**Comment by thomashoneyman** [comment]
As I look at this again -- `Free` is exported without constructors, so it might be safe for me to just directly drop in this new implementation of `Free` in the modules where it was previously used, which would cut down on the `Compat` modules quite a bit.

File: src/Control/Monad/Free.purs [hunk 1]
```diff
@@ -1,245 +1,146 @@
...
+data FreeCons f a b
+  = FreeCons (a -> Free f UnsafeBoundValue) (FreeBinds f UnsafeBoundValue b)
 
-instance ordFree :: (Functor f, Ord1 f, Ord a) => Ord (Free f a) where
-  compare x y = case resume x, resume y of
-    Left fa, Left fb -> compare1 fa fb
-    Left _, _ -> LT
-    _, Left _ -> GT
-    Right a, Right b -> compare a b
+lift :: forall f a. f a -> Free f a
```
Comment (xgrommx) [inline src/Control/Monad/Free.purs:42]: I don't like it
1) Breaking changes
2) `lift ` is standard name for `MonadTrans`

Comment (thomashoneyman) [reply src/Control/Monad/Free.purs:42]: I'm open to discussing whether `liftF` should remain the name of this function. It's not _only_ used for `MonadTrans` (it's also used for say Halogen's lift, or `run`'s lift) but if folks would really like it to remain `liftF` I can see why.

Do you have any other comments for this PR?

**Comment by safareli** [comment]
It seams that if you have:

```purs 
loop x = Free.pure (x + 1) >>= loop
oops = loop 0
```

This will blow stack before it's even interpreted

**Comment by natefaubion** [comment]
That will always happen. You are building an infinite tree strictly since there’s no eta expansion within loop.

**Comment by safareli** [comment]
Ok, but here it's eta expanded version:
```
loop x = Free.pure (x + 1) >>= \k => loop k
oops = loop 0
```

Because bind is implemented like this: `bind (Pure a) k = k a` I suspect it will blow stack, wouldn't it?

**Comment by thomashoneyman** [comment]
A couple of updates for y'all:

* In https://github.com/purescript/purescript-free/pull/104/commits/6180532bdbccdf0443a7e7beebef8d8ff5655bfe I changed the `Compat` modules so that they only preserve the prior API but re-use the new implementation. This is possible because `Free`'s constructors were never exported, so this isn't breaking. This greatly reduces the number of affected modules.
* In https://github.com/purescript/purescript-free/pull/104/commits/9377e08b4e994aa7097bb9a5cbbc72578faf5bc4 I added back the missing instances for `Eq`, `Ord`, `Eq1`, `Ord1`, `Foldable`, `Traversable`, `Semigroup`, `Monoid`, `MonadRec`, and `MonadTrans`. It's not clear to me that all of these instances are necessary, but removing any of them _would_ be a breaking change not addressable via a `Compat` module now that the same implementation is reused there.

At this point, beyond any discussion we need to have about potential issues with the implementation itself, it really comes down to whether these API changes are palatable or not. If they aren't, then I'd like to see where we can meet in the middle on some of these function renames and changes.

File: src/Control/Monad/Free.purs [hunk 2]
```diff
@@ -1,245 +1,233 @@
...
+  bind (Pure a) k = k a
+  bind (Bind a bs) k = Bind a (Node (unsafeCoerce bs) (Leaf (unsafeCoerce k)))
 
-instance freeMonad :: Monad (Free f)
+instance monadFree :: Monad (Free f)
 
-instance freeMonadTrans :: MonadTrans Free where
-  lift = liftF
+instance monadTransFree :: MonadTrans Free where
+  lift = lift
```
Comment (thomashoneyman) [inline src/Control/Monad/Free.purs:73]: @xgrommx The `lift` in this library is the same as the `lift` from `MonadTrans` -- it's one reason why I think it's sensible to rename the function.

**Comment by thomashoneyman** [comment]
Here are the API changes:

**Renamed**
These renames largely drop the `F` / `Free` / `M` suffixes, continuing the progress made when this library moved away from the many variants on `liftFree`, `liftFC`, etc. that existed prior to the current version of this library. Instead, we use `Rec` as a suffix to indicate that a function has a `MonadRec` constraint, and otherwise try to avoid suffixes.

`liftF` -> `lift`
`wrap` -> `roll`
`suspendF` -> `suspend`
`hoistFree` -> `hoist`
`substFree` -> `interpret`
`foldFree` -> `interpretRec`
`runFree` -> `run`
`runFreeM` -> `runRec`
`resume'` -> `resume`

Removed:

`resume`, because it is trivial to implement via the new `resume` (aka the old `resume`) along with any other variant you might wish for. It could be re-added under a different name.

**Comment by natefaubion** [comment]
> Because bind is implemented like this: bind (Pure a) k = k a I suspect it will blow stack, wouldn't it?

Ah, yep! That's a great point. It needs to just be suspended as a bind instead of taking a shortcut.

File: src/Control/Monad/Free/Compat.purs [hunk 3]
```diff
@@ -0,0 +1,82 @@
+module Control.Monad.Free.Compat
```
Comment (thomashoneyman) [inline src/Control/Monad/Free/Compat.purs:1]: An alternative to this module would be to keep these functions in the original module, and add deprecation notices to them ("in the next major version this function will be renamed to <x>"). Do y'all have a preference between those two approaches?

**Comment by thomashoneyman** [comment]
I have reverted changes that would break the existing API. I'm still of the opinion we *should* make those changes, but I shouldn't have pushed to make them in this PR. I've updated the PR description with details:

https://github.com/purescript/purescript-free/pull/104#issue-501219110

This PR now offers a performance improvement without breaking changes.

**Comment by thomashoneyman** [comment]
@natefaubion @safareli I've verified that this blows up (at compile time):

```purs
loop x = Free.pure (x + 1) >>= \k => loop k
oops = loop 0
```

I'm not quite sure how to suspend `Pure` as a bind; while `k` can be dealt with the same as the existing `Bind` implementation, `a` is not so simple.

Here's the existing implementation ([source](https://github.com/purescript/purescript-free/pull/104/files#diff-f92724465af909a5e70e5ce334dea9a6de10ca424c32d1bd33594d6956a73f3dR70-R72
)):

```purs
instance bindFree :: Bind (Free f) where
  bind (Pure a) k = k a
  bind (Bind a bs) k = Bind a (Node (unsafeCoerce bs) (Leaf (unsafeCoerce k)))
```

I can partially rewrite this as:

```purs
instance bindFree :: Bind (Free f) where
  bind (Pure a) k = Bind (?wat a) (Leaf (unsafeCoerce k))
  bind (Bind a bs) k = Bind a (Node (unsafeCoerce bs) (Leaf (unsafeCoerce k)))
```

producing the error

```
Hole 'wat' has the inferred type
                             
  a0 -> f1 UnsafeBoundValue
                             
You could substitute the hole with one of these values:
                                                     
  Unsafe.Coerce.unsafeCoerce  :: forall a b. a -> b
                                                     
in the following context:

  a :: a0
  k :: a0 -> Free f1 b2

in value declaration bindFree
```

---

I'm open to suggestions on how to address this, and I'll keep working on it on my own!

**Comment by safareli** [comment]
This shouldwork but it might not be as elegant 


  ```diff
= Pure a
+  | BindPure UnsafeBoundValue (FreeBinds f UnsafeBoundValue a)
   | Bind (f UnsafeBoundValue) (FreeBinds f UnsafeBoundValue a)
```

**Comment by thomashoneyman** [comment]
@safareli That's a good idea. I gave it a try and I was able to get it working except for one place -- `resume'`:

```purs
resume'
  :: forall f a r
   . (forall b. f b -> (b -> Free f a) -> r)
  -> (a -> r)
  -> Free f a
  -> r
resume' bind' pure' = case _ of
  Pure a -> pure' a
  BindPure a bs -> bind' ?a (go1 bs)
  Bind a bs -> bind' a (go1 bs)
  where
  go1 :: forall x y. FreeBinds f x y -> x -> Free f y
  go1 bs x = case bs of
    Leaf k -> k x
    Node l r -> case uncons l r of
      FreeCons k bs' -> case k x of
        Pure a -> go1 bs' a
        BindPure a bs'' -> BindPure a (Node bs'' bs')
        Bind a bs'' -> Bind a (Node bs'' bs')
    Hoist nat bs' ->
      go2 nat bs' x

  go2 :: forall g x y. (UnsafeBoundF ~> g) -> FreeBinds UnsafeBoundF x y -> x -> Free g y
  go2 nat bs x = case bs of
    Leaf k -> hoistFree nat (k x)
    Node l r -> case uncons l r of
      FreeCons k bs' -> case k x of
        Pure a -> go2 nat bs' a
        BindPure a bs'' -> BindPure a (Hoist nat (Node bs'' bs'))
        Bind a bs'' -> Bind (nat a) (Hoist nat (Node bs'' bs'))
    Hoist nat' bs' ->
      go2 (nat <<< nat') bs' x
```

That's quite a bit of code -- this is the excerpt worth looking at:

```purs
resume' bind' pure' = case _ of
  Pure a -> pure' a
  BindPure a bs -> bind' ?a (go1 bs)
  Bind a bs -> bind' a (go1 bs)
  where
  ...
```

which produces the following error with the typed hole:

```
Compiling Control.Monad.Free
Error found:
in module Control.Monad.Free
at src/Control/Monad/Free.purs:232:26 - 232:28 (line 232, column 26 - line 232, column 28)

  Hole 'a' has the inferred type
                       
    f0 UnsafeBoundValue
                       
  in the following context:

    a :: UnsafeBoundValue
    bind' :: forall b. f0 b -> (b -> ...) -> r2
    bs :: FreeBinds f0 UnsafeBoundValue a1
    go1 :: forall x y. FreeBinds f0 x y -> x -> Free f0 y
    go2 :: forall g x y. NaturalTransformation UnsafeBoundF g -> FreeBinds UnsafeBoundF x y -> x -> ...


in value declaration resume'
```

...which feels like the same issue, deeper in the code. Might just be missing something simple, though :)

**Pull request closed without merging**